src/server/middlewares/validateVote.js

```diff
@@ -1,175 +1,260 @@
+    .getVotingBooth(vote.boothAddress)
+    .call();
+  const isInitialized = result[0];
+  const isDeactivated = result[1];
+  const festivalId = result[2];
+
+  const festivalChainId = await questionContract.methods.festival().call();
+
+  // Check if ..
+  // 1. Festival is active
+  // 2. Voting booth is actually connected to this festival
+  if (!isInitialized || isDeactivated || festivalChainId !== festivalId) {
+    throw Error('Invalid booth address');
   }
-};
+}
 
-const checkChainHasVoted = async (vote, festivalQuestion, artworkQuestion) => {
-  const hasVotedFestival = await festivalQuestion.methods
-    .hasVoted(vote.sender)
+async function checkHasVoted(
+  vote,
+  festivalQuestionContract,
+  artworkQuestionContract,
+) {
+  const hasVotedFestival = await festivalQuestionContract.methods
+    .hasVoted(vote.senderAddress)
     .call();
-  const hasVotedArtwork = await artworkQuestion.methods
-    .hasVoted(vote.sender)
+
+  const hasVotedArtwork = await artworkQuestionContract.methods
+    .hasVoted(vote.senderAddress)
     .call();
+
   if (hasVotedFestival || hasVotedArtwork) {
-    throw Error('has voted');
+    throw Error('Sender voted already');
   }
-};
+}
 
-const checkLocalHasVoted = async (vote) => {
+async function checkLocalHasVoted({
+  senderAddress,
+  festivalQuestionAddress,
+  artworkQuestionAddress,
+}) {
   const localHasVotedFestival = await Vote.findOne({
-    where: { sender: vote.sender, festivalQuestion: vote.festivalQuestion },
+    where: {
+      senderAddress,
+      festivalQuestionAddress,
+    },
   });
+
   const localHasVotedArtwork = await Vote.findOne({
-    where: { sender: vote.sender, artworkQuestion: vote.artworkQuestion },
+    where: {
+      senderAddress,
+      artworkQuestionAddress,
+    },
   });
+
   if (localHasVotedFestival || localHasVotedArtwork) {
-    throw Error('has voted');
+    throw Error('Duplicate vote in database');
   }
-};
+}
 
-const checkArtworkQuestionIsHighestVoted = async (vote) => {
-  let highestVote = Array.from(new Set(vote.festivalVoteTokens));
-  highestVote = highestVote.sort((itemA, itemB) => itemB - itemA)[0];
+async function checkArtworkQuestionIsHighestVoted(vote) {
+  // Question does not exist or is not for artworks
   const question = await Question.findOne({
-    where: { address: vote.artworkQuestion },
+    where: {
+      address: vote.artworkQuestionAddress,
+    },
   });
+
   if (!question || !question.artworkId) {
-    throw Error('invalid artwork-question');
+    throw Error('Question is not for artwork answers');
   }
-  const justAnswerIds = vote.festivalAnswers.map((a) => a.id);
-  const indexOfArtwork = justAnswerIds.indexOf(question.artworkId);
-  if (vote.festivalVoteTokens[indexOfArtwork] !== highestVote) {
-    throw Error('invalid artwork-question');
+
+  // We can only vote on properties of an artwork with the highest score of the
+  // previous vote
+  const highestAnswer = vote.festivalVoteTokens
```

Comment:
If I'm reading this right, it assumes there is actually a single highest answer, when actually it might be that the most vote tokens any answer had is 3, and all we should be checking is that the answer they're voting on had 3 votes. Again, that's what's going on with the cast to set in highestVote